### PR TITLE
Initial commit for new germline workflow

### PR DESCRIPTION
**Pull request by alexiswl**
## Changes made:

* germline.py:
  * now creates workflow_input with fastq_list_rows input.   
  * input template needs only sample_name as the other input to be launched.  

* Initialised ica.py:    
  * Identical  to iap.py but with ICA_ACCESS_TOKEN and ICA_BASE_URL env vars.  
  * Have kept IAP_JWT_TOKEN and IAP_BASE_URL constants as fallbacks for now.  

* Initialised libgds.py :    
  * Shortcuts for downloading a file as a tempfile object.  
 
* requirements.txt:  
  * Added in libica into the requirements.

### CODE REVIEW

**Comment by alexiswl** [comment]
## TODO list

- [ ] Update input ssm parameter template object
- [ ] Update workflow id ssm parameter
- [ ] Update workflow version ssm parameter

**Comment by victorskl** [comment]
> ## TODO list
> * [ ]  Update input ssm parameter template object
> * [ ]  Update workflow id ssm parameter
> * [ ]  Update workflow version ssm parameter
> 

For this, you can use data portal terraform sub-stack called [workflow](https://github.com/umccr/infrastructure/blob/master/terraform/stacks/umccr_data_portal/workflow/main.tf)  here.

tl;dr
```
cd infrastructure
cd terraform/stacks/umccr_data_portal/workflow
yawsso -p dev
export AWS_PROFILE=dev
terraform workspace select dev
(do `terraform init` if it is fresh start)
terraform plan
```

Make changes in `main.tf` accordingly, then
```
terraform plan
terraform apply
```

**Review by victorskl — COMMENTED** [review]
LGTM for impl-wise. Need fix/update related test cases.

Since code logic impl has changed, please update related unit/IT test cases (i.e align up fixtures/mock) accordingly.

tl;dr

to run smoke test includes every aspect of data portal
```
aws sso login --profile dev && export AWS_PROFILE=dev
yawsso -p dev

make up
docker ps

make test
```

- to run all tests related to ICA pipeline package
```
python manage.py test data_processors.pipeline.tests
```

- to run all tests related to germline module
```
python manage.py test data_processors.pipeline.tests.test_germline
```

- see each test case pydoc string for how to invoke individually

Happy to do help pair programming run for this, if any.

File: utils/libgds.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+# Standard imports
+import logging
+import requests
+from tempfile import NamedTemporaryFile
+from datetime import datetime
+
+# ICA imports
```
Comment (victorskl) [inline utils/libgds.py:7]: Ah. Just suggestion, I would name this util module as `gds.py` (which denote app aspect of it) and, to avoid confusion with SDK library naming.

**Comment by alexiswl** [comment]
> LGTM for impl-wise. Need fix/update related test cases.
> 
> Since code logic impl has changed, please update related unit/IT test cases (i.e align up fixtures/mock) accordingly.
> 
> tl;dr
> 
> to run smoke test includes every aspect of data portal
> 
> ```
> aws sso login --profile dev && export AWS_PROFILE=dev
> yawsso -p dev
> 
> make up
> docker ps
> 
> make test
> ```
> 
> * to run all tests related to ICA pipeline package
> 
> ```
> python manage.py test data_processors.pipeline.tests
> ```
> 
> * to run all tests related to germline module
> 
> ```
> python manage.py test data_processors.pipeline.tests.test_germline
> ```
> 
> * see each test case pydoc string for how to invoke individually
> 
> Happy to do help pair programming run for this, if any.

We haven't touched the inputs to the context yet (though this might change once we move to the bclconvert 3.7.5 workflow and store the fastq_list_rows output).  Therefore, having a look at the tests, I don't think any of the tests need changing since it still expects the same inputs.

File: utils/libgds.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    From a gds path, collect the file id.
+    One can then use the file id to collect other metadata on a file
+    """
+
+    configuration = libgds.Configuration(
+        host="https://aps2.platform.illumina.com",
+        api_key={
+            'Authorization': 'YOUR_API_KEY'
+        }
+    )
```
Comment (reisingerf) [inline utils/libgds.py:29]: This is not used? Meant as comment?

File: utils/libgds.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    :param gds_path: the GDS path of the file to download
+    :return local_path: or None if file not found
+    """
+
+    logger.info(f"Downloading file from GDS: gds://{gds_volume_name}{gds_path}")
+
+    gds_file_id = get_gds_file_id_from_gds_path(volume_name=gds_volume_name,
+                                                path=gds_path)
+
+    gds_file_obj = get_file_obj_from_gds_file_id(gds_file_id)
```
Comment (reisingerf) [inline utils/libgds.py:102]: Does this not retrieve the same information twice?

**Review by reisingerf — COMMENTED** [review]
Looks good. Inline comments

File: utils/libgds.py (hunk 3 above)
Comment (alexiswl) [reply utils/libgds.py:102]: Yes, essentially would like to be able to run a check that the file exists and then retrieve the file object, which is the objective of the first function.  The second function can then be used in any case where one has a file id.

File: utils/libgds.py (hunk 2 above)
Comment (alexiswl) [reply utils/libgds.py]: Resolved in 495ec2001399f8b6125a8b4b6e3cd544794d3d82

**Comment by alexiswl** [comment]
> > ## TODO list
> > 
> > * [ ]   Update input ssm parameter template object
> > * [ ]   Update workflow id ssm parameter
> > * [ ]   Update workflow version ssm parameter
> 
> For this, you can use data portal terraform sub-stack called [workflow](https://github.com/umccr/infrastructure/blob/master/terraform/stacks/umccr_data_portal/workflow/main.tf) here.
> 
> tl;dr
> 
> ```
> cd infrastructure
> cd terraform/stacks/umccr_data_portal/workflow
> yawsso -p dev
> export AWS_PROFILE=dev
> terraform workspace select dev
> (do `terraform init` if it is fresh start)
> terraform plan
> ```
> 
> Make changes in `main.tf` accordingly, then
> 
> ```
> terraform plan
> terraform apply
> ```

I've created a pull request [here](https://github.com/umccr/infrastructure/pull/112).  Waiting on WES release to fix https://github.com/umccr-illumina/cwl-iap/issues/184 so we can test this out on dev and then push to prod

File: utils/libgds.py (hunk 3 above)
Comment (alexiswl) [reply utils/libgds.py]: Do you have a better way to do this? There is no equivalent of `ica files get gds://path/to/file` on the API side? One **HAS** to use a file id in order to retrieve information about the file.

**Comment by victorskl** [comment]
> I've created a pull request [here](https://github.com/umccr/infrastructure/pull/112) and then re-run `terraform plan` with the outputs shown below. Waiting on WES release to fix [umccr-illumina/cwl-iap#184](https://github.com/umccr-illumina/cwl-iap/issues/184) so we can test this out on dev and then push to prod.

Okay, sound good! Please feel free to apply to `dev` and test there.

P.S. you can use no color output flag like `terraform plan -no-color | pbcopy`

File: utils/libgds.py (hunk 3 above)
Comment (reisingerf) [reply utils/libgds.py]: Use only the list_files request? (adding the `presignedUrl` include)
And perhaps instead of returning only the file ID(s), return the full objects?

Comment (alexiswl) [reply utils/libgds.py]: Resolved in 6611457c3cba686ba526a58f4e6b86c6b9276fda

**Comment by alexiswl** [comment]
@victorskl added a few #FIXMEs in 6611457c3cba686ba526a58f4e6b86c6b9276fda.  Rather than returning None I should be raising errors instead.  Do we have any custom errors for "GDSFileNotFound" in this repo?

**Comment by victorskl** [comment]
> Do we have any custom errors for "GDSFileNotFound" in this repo?

Nop. Feel free to create one, if any.

**Comment by alexiswl** [comment]
Superceded by #157
